src/blob/index/bptree/core.rs

```diff
@@ -0,0 +1,429 @@
+/// structure of b+-tree index file from the beginning:
+/// 1. Header
+/// 2. Bloom filter
+/// 3. Sorted by keys array of record headers (serialized size of one record header is 61 bytes)
+/// 4. Metadata
+/// 5. Sorted by keys array of pairs `key - offset of record header in file` (these are leaf nodes
+///    of our B+ tree; as opposed to using record headers as leaves, this approach allows to include
+///    more elements in one block and reducing height of the tree (but increasing on one blocks
+///    that we need to read); this also will be useful for scanning queries: instead of scanning
+///    61 bytes headers (in case of 4-bytes key) we scan 8 bytes pairs (in case 4-bytes key and
+///    4-bytes offset) which allows to read less pages in almost 8 times)
+/// 6. Non-leaf nodes of btree in format
+///    `NodeMeta - pointer - key - pointer - key - ... - pointer`,
+///    where pointer - is offset in file on underlying node with searched key
+use super::prelude::*;
+
+const BLOCK_SIZE: usize = 4096;
+
+#[derive(Debug, Clone)]
+pub(crate) struct BPTreeFileIndex {
+    file: File,
+    header: IndexHeader,
+    metadata: TreeMeta,
+}
+
+#[async_trait::async_trait]
+impl FileIndexTrait for BPTreeFileIndex {
+    async fn from_file(name: FileName, ioring: Option<Rio>) -> Result<Self> {
+        trace!("open index file");
+        let file = File::open(name.to_path(), ioring)
+            .await
+            .context(format!("failed to open index file: {}", name))?;
+        let header = Self::read_index_header(&file).await?;
+        let metadata = Self::read_tree_meta(&file, &header).await?;
+
+        Ok(Self {
+            file,
+            header,
+            metadata,
+        })
+    }
+
+    async fn from_records(
+        path: &Path,
+        ioring: Option<Rio>,
+        headers: &InMemoryIndex,
+        filter: &Bloom,
+    ) -> Result<Self> {
+        let res = Self::serialize(headers, filter)?;
+        if res.is_none() {
+            error!("Indices are empty!");
+            return Err(anyhow!("empty in-memory indices".to_string()));
+        }
+        let (mut header, metadata, buf) = res.expect("None case is checked");
+        let _ = std::fs::remove_file(path);
+        let file = File::create(path, ioring)
+            .await
+            .with_context(|| format!("file open failed {:?}", path))?;
+        file.write_append(&buf).await?;
+        header.written = 1;
+        let serialized_header = serialize(&header)?;
+        file.write_at(0, &serialized_header).await?;
+        file.fsyncdata().await?;
```

Comment:
In my opinion we do should fsync when we create file index from records because:
1. If higher-level struct requested creation `from_records` (not `from_file`) then some data changed => then we need to rewrite file (=> then we need to fix state (read (2))
2. Once data is on disk, it's read-only (write operations are not permitted), so if we fsync at the start we don't need to care about when we should do it and potential chance to lose all data  in case of unexpected disk dropping out or something like that (remember about page cache: the data in index file without fsync will likely be corrupted because of dirty pages in RAM, that are not synced with disk)

BUT! I agree that it wouldn't be redundant to add flag in `IndexStruct`: if there were any successful write operations since creating InMemoryIndex - we should create file index from records, otherwise - we can create it from file!
What do you think?

Reply:
I need to review index life-cycle one more time. I will come back later this week. @pyakushin Could you check it as well?

Reply:
@pyakushin I mean data write in active blob. During this operation `InMemoryIndex` will be updated (and will have at least 1 record header) so it can be successfully serialized. But if it hasn't at least one record: 
1. we are not able to determine key size; 
2. it's not worth to store file without record headers at all.

Reply:
On the first part: Yes, I meant `flush` operation. It is only applicable for `index` files.
> if dump is processed once a day, shouldn't we optimize this by doing this non-simultanously instead of changing fsync on flush?

It would be good optimization. How hard to do this?

On the second part: ok, I get it. This is a good optimization task, but I think it is better to do it in a separate request.

Reply:
Change fsync on flush for headers.
> It would be good optimization. How hard to do this?
Pearl is only responsible for closing blobs on bytes/records limit excess. 

Bob is responsible for closing blob due to timestamp excess, so this work should be done in Bob crate (it doesn't looks like a difficult task: it would be enough to append offset to timestamp start (and end) value(-s) according, for example, to vdiskid)